### DISCUSSION TITLE
Command palette

### DISCUSSION THREAD

**Discussion by geekscrapy** [saulpw/visidata#2113] [❤️]
Just wanted to drop a note to say: "I bloody love the command palette" :D

A very welcome inclusion, and makes entering the commands a breeze. I find myself often spelling the command _extremely_ lazily, missing bits off when I know there are a limited amount of commands with that name (e.g. I can enter `sycopc` for `syscopy-cells` - this may be slightly incorrect, doing it from memory, but you get the gist).

My only things to nitpick is to change the colors of the identified letters to a stronger color, or maybe bold (as well?). Also reducing the size of the green palette background to the number of commands on show, I find my eyes drawn to the top of the green palette even when there is only one command highlighted.

But, loving it, thanks!

**Comment by saulpw** [comment]
Thanks so much, @geekscrapy!  Do you use the number keys then to select a command?  Do you learn the "actual" keybinding then, or do you find yourself developing idiosyncratic usage like `Space syp 2`?

**Reply by geekscrapy** [reply]
I use the number key once I see what I want - and have got very ~~lazy~~ quick at typing the full commands because of it.

I find myself thinking (and typing):
I want to copy to the system clipboard: types "sys"
I want to copy: types "cop"

I am definitely (in a good way!) forgetting the specific long commands.

There was one other improvement that I forgot: The ability to press up when entering the command to select the one you want (so it highlights the commands row). At the moment I find it a multi-step process: enter command, look at suggestions, _lookup the commands number_, find number on keyboard. Where-as I often find myself (still after about of week of using it) reaching for the arrow keys to select the command. I have a lot of muscle memory with combo-boxes, and the new vd menu is keyboard selectable it might make sense to add this ability.

FWIW I have large fingers (I dont, just can't type reliably) I will often type the shortcut differently and therefore the command has a different number.

**Reply by saulpw** [reply]
So, the up-arrow would interfere with command history, but I like this idea since the current number selection scheme interferes with being able to type e.g. `q4` in the aggregator selection to get quartiles.  Maybe this is where @frosencrantz's suggestion for `Ctrl+R` reverse history search would be useful instead of the current history mechanism?  But I do like using `/ Up` to get my previous search back like in vim.

**Reply by geekscrapy** [reply]
I hadn't thought about the `q4` etc. selection. oops!

Yea, I think the ctr+r would be best, I admit, it has popped into my head a couple of times while wanting to find a command, I think it would fit nicely here though, it'd still be fuzzy matching.

Also, while we are thinking about input mods; is ctrl+del/backspace supposed to work (as in delete a word back)?
E.g.
- typed: syscopy-cell
- press: ctrl+backspace
- prompt shows: syscopy-

**Reply by saulpw** [reply]
Hmmm...`Ctrl+Del` results in curses `kDC5`, so we can make that work, but `Ctrl+Backspace` seems to be the same as `Backspace` or `Ctrl+H` on my setup, so we can't differentiate.  And yes, if there's something that you expect to work, but doesn't, but we can make it work, then if it's not too complicated, we should!  So it sounds like you expect `Ctrl+Del` to be the same as `Ctrl+W` in input mode?

**Reply by daviewales** [reply]
Late to the party, but consider how Vim and Neovim deal with the problem of navigating both history and completion.

Suppose I type `:he`.
If I now press the 'up arrow', I will get the last command I typed starting with `:he`.
However, if I press `Tab`, I cycle through a completion menu. If I want to get directly to the command I want, I need to type a unique set of characters, then press `Tab`. (If I have already pressed `Tab`, I need to start again, or just press `Tab` until I get to where I'm going.)

Rather than numeric selection to run a suggested command, why not just keep typing until the command you want is at the top of the suggestion list, then `Tab`, `Enter`? (This already works, and I'm more likely to do this than hunt down the list to figure out what number to press...)

**Comment by anjakefala** [comment]
Btw, @geekscrapy, the colours of the identified letters are set by `vd.theme_option('color_match', 'red', 'color for matching chars in palette chooser')`. You can play with changing the colour, and show us screenshots of options that you prefer!

**Reply by geekscrapy** [reply]
Yea I saw this option after a quick hunt. The red colour doesn't seem to be working for me on iterm on Mac. I'll check on it again. Red works in other places though

**Reply by geekscrapy** [reply]
![image](https://github.com/saulpw/visidata/assets/11225502/478a0b7f-dccf-4588-9f90-65552e2162da)
This is what is shown by default :|

**Reply by saulpw** [reply]
Where does red work for you, and what does it look like?

**Reply by geekscrapy** [reply]
<img width="686" alt="image" src="https://github.com/saulpw/visidata/assets/11225502/50f5363a-6839-421a-a097-2630336a76c3">

I currently use it in the col (I was going to show the whole interface as red but I cant see the expected options now: i.e. color_menu), is there a new flag to show this?

**Reply by saulpw** [reply]
it's unless options.disp_help < 0.  I'm not sure about that whole system, I think some options are "advanced" but I'm not sure how advanced.

**Reply by maxim-uvarov** [reply]
+1, red color is very inconvenient for me on my Mac too (wezterm - default dark color scheme, if it is important). 

P.S. I also want to express so much love for the Command Palette! It has given me a huge performance and friendliness boost!

**Comment by geekscrapy** [comment]
Ideas from the discord chat after using it for a couple of days

I think the way I saw it working:

(When I say "closest", I mean from the input prompt row. Therefore if I'm typing into the input field and press UP arrow, the closest next row is the one directly above the input field row)

Input start of command - this brings up commands matching the input. Closest match at the bottom of the pallet. This closest match is *pre-selected*, therefore if you press enter it runs it automatically.

Having the closest match at the bottom *visually* joins the input field and the suggestions, and you can quickly compare. In addition when you press UP to interact with the pallet, it feels more joined than jumping to the top of another panel with tab.

From here if you press UP it would cycle through all the other close options and scroll up if required (I have a few custom commands which have the same prefix). The commands are always completed in the input (when focus is moved to the pallet), so what is selected can always been run by just hitting ENTER.

If there are two matching commands for the input, there should be 2 lines of suggestions?

I think the only time that recent commands should be shown is when there is no input in the input field, and you can then access these by just pressing UP. (Most recent at the bottom of the pallet)

If there are multiple commands matching the prefix in the input field, the order of the closest command is dictated by it's order in history, more recently used == quicker to select.

This is just some aspects of how I envisioned it after playing for a bit as I seem to be a tad confused by it at the moment 😬

It seems the key presses I have to do to select the 2nd suggestion is:

1. Type command in input field
2. TAB - Makes the numbers show in the pallet
3. Press the number 2

If I TAB TAB ENTER, it has the 2nd option selected (👍), but attempts to run whatever half input command is in the input
